### ISSUE TITLE
Multi Timeline Nonces for out of order execution

### ISSUE THREAD

**Issue by CJ42** [lukso-network/lsp-smart-contracts#15]
This issue follows discussion with @frozeman and @YamenMerhi.

## What this issue is about?

We want relayer services to be able to send relayed transactions, independently of the order of execution of nonces.

For instance, if executing 3 transactions that send 1 LYX to Alice, Bob and Carol, Carol should not have to wait for Alice and Bob transaction to complete to receive its token.

Each transactions should be executed on separate **timelines**

- Relayer -> Alice tx -> Account   => timeline 1
- Relayer -> Bob tx -> Account    => timeline 2
- Relayer -> Carol tx -> Account  => timeline 3

## Multi timeline nonces

Based on Permit Contract: https://github.com/Amxx/Permit

```
==========================================================================>
(timeline 0) nonce   0 --> 1 --> 2 --> 3   
(timeline 1)         0
(timeline 2)         0
(timeline 3)         0
(timeline n)         0
(...)
```

[Unlike in this example of use of multi timeline nonce, it is not necessary to use `uint128` bit shifting to represent the timeline.](https://github.com/Amxx/Permit/blob/9429ab2233c27cd6bb42718d0dda829c5df95352/contracts/EIP712WithNonce.sol#L23)

Nonces can just be represented in a nested mapping. The second key represents the **timeline**. where:
- timeline 0 = sequential order
- timeline 1 or more = out of order

First time we use a `timelineN`, gas cost = 20000 (write to slot for `timelineN` for the first time) + execution cost.
If a `timelineN` is reused, gas cost = 5000 (update slot for `timelineN`) + execution slot.

1. relay service calls `getNonce(signer, timeline)`) -> return nonce (for `timeline`)

2. `executeRelay(...)` function should be overriden like this:

```solidity
function executeRelayCall(
        // other parameters...
        uint256 _nonce  // retrieved from 1.
        uint256 _timeline // 0 for sequential, 1 or more for out of order
        bytes memory _signature
    ) external payable returns (bool success_) {

        // checks signedFor is the KeyManager + create blob to recover address

        // recover the signer
        address from = keccak256(blob).toEthSignedMessageHash().recover(
            _signature
        );

        require(
            _nonceStore[from][_timeline] == _nonce,
            "KeyManager:executeRelayedCall: Incorrect nonce"
        );

        // increase the nonce
        _nonceStore[from][_timeline] = _nonceStore[from][_timeline].add(1);

        // check permissions + make call
    }
```

We could use the `batchRequest` from web3 to test this: https://web3js.readthedocs.io/en/v1.4.0/web3-eth.html?highlight=batch#batchrequest

**Comment by CJ42** [comment]
@frozeman @YamenMerhi 
Feature implemented in #17 

https://github.com/lukso-network/lsp-universalprofile-smart-contracts/pull/17

**Comment by YamenMerhi** [comment]
Description of Multi-channel nonce is found here: https://docs.lukso.tech/standards/faq/channel-nonce
